File: cli/services/lightrag.py
"""LightRAG route wrappers."""

from __future__ import annotations

from typing import Any
from urllib.parse import urlencode

from cli.api_client import ApiClient


class LightRagService:
    def __init__(self, client: ApiClient):
        self._client = client

    def list_labels(self, domain_id: str = "default") -> list[dict[str, Any]] | list[str]:
        payload = self._client.get(f"/lightrag/domains/{domain_id}/graph/labels")
        return payload if isinstance(payload, list) else []

    def popular_labels(
        self,
        *,
        domain_id: str = "default",
        limit: int = 20,
    ) -> list[dict[str, Any]] | list[str]:
        payload = self._client.get(
            f"/lightrag/domains/{domain_id}/graph/labels/popular?{urlencode({'limit': limit})}"
        )
        return payload if isinstance(payload, list) else []

    def search_labels(
        self,
        *,
        domain_id: str = "default",
        query: str,
        limit: int = 20,
    ) -> list[dict[str, Any]] | list[str]:
        payload = self._client.get(
            f"/lightrag/domains/{domain_id}/graph/labels/search?{urlencode({'q': query, 'limit': limit})}"
        )
        return payload if isinstance(payload, list) else []

    def get_graph(
        self,
        *,
        domain_id: str = "default",
        label: str,
        max_depth: int = 3,
        max_nodes: int = 1000,
    ) -> dict[str, Any]:
        payload = self._client.get(
            f"/lightrag/domains/{domain_id}/graphs?"
            f"{urlencode({'label': label, 'max_depth': max_depth, 'max_nodes': max_nodes})}"
        )
        return payload if isinstance(payload, dict) else {}

File: client/src/api/lightrag.ts
import { apiRequest } from "@/lib/api/client";
import { popularLabelsDefaultLimit, searchLabelsDefaultLimit } from "@/lib/constants";
import { getSelectedLightRagDomainId } from "@/stores/lightrag-domain-store";

export type LightragNodeType = {
  id: string;
  labels: string[];
  properties: Record<string, unknown>;
};

export type LightragEdgeType = {
  id: string;
  source: string;
  target: string;
  type?: string;
  relation?: string;
  weight?: number;
  description?: string;
  properties: Record<string, unknown>;
};

export type LightragGraphType = {
  nodes: LightragNodeType[];
  edges: LightragEdgeType[];
  truncated?: boolean;
};

export type EntityUpdateResponse = {
  status: string;
  message: string;
  data: Record<string, unknown>;
  operation_summary?: Record<string, unknown>;
};

export type DocActionResponse = {
  status: string;
  message: string;
};

export type QueryMode = "naive" | "local" | "global" | "hybrid" | "mix" | "bypass";

export type Message = {
  role: "user" | "assistant" | "system";
  content: string;
};

export type QueryRequest = {
  query: string;
  mode: QueryMode;
  only_need_context?: boolean;
  only_need_prompt?: boolean;
  response_type?: string;
  stream?: boolean;
  top_k?: number;
  chunk_top_k?: number;
  max_entity_tokens?: number;
  max_relation_tokens?: number;
  max_total_tokens?: number;
  conversation_history?: Message[];
  history_turns?: number;
  user_prompt?: string;
  enable_rerank?: boolean;
};

export async function queryGraphs(label: string, maxDepth: number, maxNodes: number): Promise<LightragGraphType> {
  const params = new URLSearchParams({
    label,
    max_depth: String(maxDepth),
    max_nodes: String(maxNodes),
  });
  return request<LightragGraphType>(`/graphs?${params.toString()}`);
}

export async function fetchGraphEntityTypes(): Promise<string[]> {
  return [];
}

export async function getPopularLabels(limit: number = popularLabelsDefaultLimit): Promise<string[]> {
  const response = await request<{ labels: string[] }>(`/graph/labels/popular?limit=${limit}`);
  return response.labels;
}

export async function searchLabels(query: string, limit: number = searchLabelsDefaultLimit): Promise<string[]> {
  const params = new URLSearchParams({ q: query, limit: String(limit) });
  const response = await request<{ labels: string[] }>(`/graph/labels/search?${params.toString()}`);
  return response.labels;
}

export async function checkEntityNameExists(entityName: string): Promise<boolean> {
  const params = new URLSearchParams({ name: entityName });
  const response = await request<{ exists: boolean }>(`/graph/entity/exists?${params.toString()}`);
  return response.exists;
}

export async function updateEntity(
  entityName: string,
  updatedData: Record<string, unknown>,
  allowRename = false,
  allowMerge = false,
): Promise<EntityUpdateResponse> {
  return request<EntityUpdateResponse>("/graph/entity/edit", {
    method: "POST",
    body: JSON.stringify({
      entity_name: entityName,
      updated_data: updatedData,
      allow_rename: allowRename,
      allow_merge: allowMerge,
    }),
  });
}

export async function updateRelation(
  sourceEntity: string,
  targetEntity: string,
  updatedData: Record<string, unknown>,
): Promise<DocActionResponse> {
  return request<DocActionResponse>("/graph/relation/edit", {
    method: "POST",
    body: JSON.stringify({
      source_id: sourceEntity,
      target_id: targetEntity,
      updated_data: updatedData,
    }),
  });
}

async function request<T>(path: string, init: RequestInit = {}): Promise<T> {
  const domainId = getSelectedLightRagDomainId();
  return apiRequest<T>(`/lightrag/domains/${encodeURIComponent(domainId)}${path}`, init);
}
